### PR TITLE
feat: add shared types

### LINKED ISSUE #66: feat: add end-to-end types, move all shared types to a shared types folder
Since nestjs doesn't have any example adapters with trpc, we currently have no type safety when querying our own backend API and rely on our own knowledge of our api types. `ts-rest` is a library that adds type safety when fetching from our nestjs backend. link: https://ts-rest.com/docs/core/

Adding end-to-end types with `ts-rest` will also involve moving all our shared types into a shared folder

### PR DESCRIPTION
**Pull request by aidanaden**
closes #66, #73

changes include:
- moves shared types between `frontend` and `backend` to the `shared` folder
- adds `CANCEL_MATCH` event where match is cancelled if either user does not accept the match/join the room after being matched (if one of the users joined the room prior to the match being cancelled, they'll be disconnected)
- adds `build` to github CI

todo:
- [x] add question API response types to `shared` folder

### CODE REVIEW

**Review by ngjunkang — APPROVED** [review]
tested, and no regression detected.

File: shared/src/api/question/schema.ts [hunk 1]
```diff
@@ -1,41 +1,8 @@
...
-export const EditableSchema = UserModel.pick({
-  email: true,
-  username: true,
-  hashRt: true,
-}).partial();
-
-export const ChangePasswordInfoSchema = z.object({
-  currentPassword: passwordZodString,
-  newPassword: passwordZodString,
+const SummaryQuerySchema = z.object({
```
Comment (blaketingyu) [inline shared/src/api/question/schema.ts:3]: lint is giving a warning that `SummaryQuerySchema` is never used

Comment (aidanaden) [reply shared/src/api/question/schema.ts:3]: can ignore, will be using later on

**Review by blaketingyu — APPROVED** [review]
LGTM

**Review by yusufaine — APPROVED** [review]
LGTM, only nit is that theres a need to manually `yarn install` in the shared folder.

**Review by yusufaine — APPROVED** [review]
LGTM!

**Pull request merged**